src/main/java/org/opensearch/knn/index/codec/jvector/JVectorGraphIndexWriter.java

```diff
@@ -0,0 +1,277 @@
+        if (graph instanceof OnHeapGraphIndex) {
+            var ohgi = (OnHeapGraphIndex) graph;
+            if (ohgi.getDeletedNodes().cardinality() > 0) {
+                throw new IllegalArgumentException("Run builder.cleanup() before writing the graph");
+            }
+        }
+        for (var featureId : featureStateSuppliers.keySet()) {
+            if (!featureMap.containsKey(featureId)) {
+                throw new IllegalArgumentException(String.format("Feature %s not configured for index", featureId));
+            }
+        }
+        if (ordinalMapper.maxOrdinal() < graph.size(0) - 1) {
+            var msg = String.format(
+                "Ordinal mapper from [0..%d] does not cover all nodes in the graph of size %d",
+                ordinalMapper.maxOrdinal(),
+                graph.size(0)
+            );
+            throw new IllegalStateException(msg);
+        }
+
+        // writeHeader(); // sets position to start writing features
+
+        // for each graph node, write the associated features, followed by its neighbors at L0
+        for (int newOrdinal = 0; newOrdinal <= ordinalMapper.maxOrdinal(); newOrdinal++) {
+            var originalOrdinal = ordinalMapper.newToOld(newOrdinal);
+
+            // if no node exists with the given ordinal, write a placeholder
+            if (originalOrdinal == OrdinalMapper.OMITTED) {
+                outData.writeInt(-1);
+                for (var feature : inlineFeatures) {
+                    throw new UnsupportedOperationException("Separated features are not supported in this implementation");
+                    // outData.seek(outData.position() + feature.featureSize());
+                }
+                outData.writeInt(0);
+                for (int n = 0; n < graph.getDegree(0); n++) {
+                    outData.writeInt(-1);
+                }
+                continue;
+            }
+
+            if (!graph.containsNode(originalOrdinal)) {
+                var msg = String.format("Ordinal mapper mapped new ordinal %s to non-existing node %s", newOrdinal, originalOrdinal);
+                throw new IllegalStateException(msg);
+            }
+            outData.writeInt(newOrdinal); // unnecessary, but a reasonable sanity check
+            assert outData.getFilePointer() == featureOffsetForOrdinal(newOrdinal) : String.format(
+                "%d != %d",
+                outData.getFilePointer(),
+                featureOffsetForOrdinal(newOrdinal)
+            );
+            for (var feature : inlineFeatures) {
+                var supplier = featureStateSuppliers.get(feature.id());
+                if (supplier == null) {
+                    throw new IllegalStateException("Supplier for feature " + feature.id() + " not found");
+                    // outData.seek(outData.position() + feature.featureSize());
+                } else {
+                    feature.writeInline(new JVectorRandomAccessWriter(outData), supplier.apply(originalOrdinal));
```

Comment:
Is writeInline different than the Feature.Inline?  If so would other feature suppliers not work?